### PR TITLE
Ninject for Xamarin.Forms

### LINKED ISSUE #279: [XF] Add Prism.Ninject.Forms
Oren Novotny released the Portable.NInject NuGet that supports Xamarin.

- Add a Prism.Ninject.Forms project and test project.
- Add to the Prism.Ninject nuget and adjust the dependencies accordingly.

I am giving this a try. Interested in a pull request if it works out?

### PR DESCRIPTION
**Pull request by rprouse**
~~**Not ready to merge**~~

Putting this up for review and possibly some ideas.

I have added a Prism.Ninject.Forms project and sample projects that use this project. I have also update the Ninject nuspec file. Everything is working as expected and behaves the same as the Unity HelloWorld application.

I added a series of unit tests for this project, but I was never able to get them to run because Xamarin.Forms needed to be Init'd before you can create an application object and Xamarin uses concrete implementations, not interfaces, so I was not able to get it set up. It looks like you ran into the same issues, thus the lack of tests in the Prism.Forms.Tests project and the fact that there is no Prism.Unity.Forms.Tests project.

Instead, I did all of my testing using the HelloWorld.Ninject project. Less than ideal, but... :frowning: 

~~The only thing that I haven't figured out yet is the `DependencyServiceExtension`. I don't know Unity well, but my read on it is that it allows Unity to resolve objects that are managed by the Xamarin.Forms DependencyService. I am not a Ninject power user, so I haven't figured out how to do the same thing in Ninject yet. Any hints on Google searches would be appreciated :smile:~~

~~I also wasn't able to test the creation of the NuGet package. It requires a signed release build and I don't have a PFX file. I will do that, I just need to go through the steps of creating a self-signed certificate.~~

Fixes #279

### CODE REVIEW

**Comment by rprouse** [comment]
My current approach is to derive a new kernel from StandardKernel that resolves from DependencyService if the requested service is not available in the kernel. Does that sound reasonable? Is there a better way?

**Comment by brianlagunas** [comment]
I am not familiar with Ninject, so I am not sure of the best way to accomplish this.  It would be good if we can find someone familiar with Ninject to help guide us on this issue.

**Comment by rprouse** [comment]
Extending the Kernel does not work. I've been reading through the Ninject source code and that of several extensions, but I don't see how I can extend Ninject to also resolve from DependencyService. I have posted to the Ninject mailing list asking for pointers, hopefully that brings a response.

Worst case, Prism for Ninject won't support constructor injection for interfaces managed by DependencyService and we document that. It may not be the only IoC container that won't support it.

I will circle back to this PR once I hear back from the mailing list. Whatever the outcome, I will clean up the code and get it ready to merge.

**Comment by rprouse** [comment]
I figured it out. They key was the interface `IMissingBindingResolver`. Once I found that, everything fell into place. This code is now functionally equivalent to the Unity version and I believe it is **ready to merge**

I don't think this needs much documentation. It works exactly the same as the Unity version except you derive from NinjectBootstrapper and it has a Kernel instead of a Container.

To test the use of DependencyService with Unity and Ninject, I added to the HelloWorld project to register platform specific classes using DependencyService and injected the interface into the constructor for the main view model and then used it in the UI.

I will go through the changes and add comments to help with the code review if any are necessary. If you have any questions or want additional changes, just let me know.

File: Samples/Xamarin/HelloWorld/HelloWorld.Ninject/HelloWorld.Ninject/App.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿using Xamarin.Forms;
+
+namespace HelloWorld
+{
+    public class App : Application
+    {
+        public App()
+        {
+            Bootstrapper bs = new Bootstrapper();
```
Comment (rprouse) [inline Samples/Xamarin/HelloWorld/HelloWorld.Ninject/HelloWorld.Ninject/App.cs]: Everything above this is just standard File > New > Xamarin Forms App. You can start code reviewing here :smile:

File: Samples/Xamarin/HelloWorld/HelloWorld.Ninject/HelloWorld.Ninject/Bootstrapper.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        protected override Page CreateMainPage()
+        {
+            return Kernel.Get<MainPage>();
+        }
+
+        protected override void RegisterTypes()
+        {
+            Kernel.RegisterTypeForNavigation<ViewB>();
+            Kernel.RegisterTypeForNavigation<ViewA>();
```
Comment (rprouse) [inline Samples/Xamarin/HelloWorld/HelloWorld.Ninject/HelloWorld.Ninject/Bootstrapper.cs]: This bootstrapper is pretty much the same as the Unity version except it derives from NinjectBootstrapper and uses a Kernel in the Ninject fashion.

File: Samples/Xamarin/HelloWorld/HelloWorld/HelloWorld.Droid/Services/MessageService.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+using HelloWorld.Interfaces;
+
+[assembly: Xamarin.Forms.Dependency(typeof(MessageService))]
+namespace HelloWorld.Droid.Services
+{
+    public class MessageService : IMessageService
+    {
+        public string Message
+        {
+            get { return "Hello from Android"; }
```
Comment (rprouse) [inline Samples/Xamarin/HelloWorld/HelloWorld/HelloWorld.Droid/Services/MessageService.cs]: One of the three platform specific classes to test interop with DependencyService. On the first screen of the app, you should see the platform specific message above the button. Tested with Unity and Ninject.

File: Samples/Xamarin/HelloWorld/HelloWorld/HelloWorld.Droid/HelloWorld.Droid.csproj [hunk 4]
```diff
@@ -22,7 +22,7 @@
     <MandroidI18n />
     <JavaMaximumHeapSize />
     <JavaOptions />
-    <TargetFrameworkVersion>v5.0</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v6.0</TargetFrameworkVersion>
```
Comment (rprouse) [inline Samples/Xamarin/HelloWorld/HelloWorld/HelloWorld.Droid/HelloWorld.Droid.csproj]: I think this is caused by an update in Xamarin Studio. The packages file below also switched from monodroid5 to monodroid6.

File: Samples/Xamarin/HelloWorld/HelloWorld/HelloWorld/ViewModels/MainPageViewModel.cs [hunk 5]
```diff
@@ -15,11 +17,14 @@ public string Title
             set { SetProperty(ref _title, value); }
         }
 
+        public string Message { get { return _messageService.Message; } }
+
         public DelegateCommand NavigateCommand { get; set; }
 
-        public MainPageViewModel(INavigationService navigationService)
+        public MainPageViewModel(INavigationService navigationService, IMessageService messageService)
```
Comment (rprouse) [inline Samples/Xamarin/HelloWorld/HelloWorld/HelloWorld/ViewModels/MainPageViewModel.cs]: `INavigationService` comes from Unity or Ninject, `IMessageService` comes from Xamarin.Forms.DependencyService.

File: Source/Xamarin/Prism.Ninject.Forms/Extensions/DependencyServiceBindingResolver.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+using Ninject.Components;
+
+using Binding = Ninject.Planning.Bindings.Binding;
+
+namespace Prism.Ninject.Extensions
+{
+    /// <summary>
+    /// Resolves missing dependencies using <see cref="Xamarin.Forms.DependencyService"/>
+    /// </summary>
+    public class DependencyServiceBindingResolver : NinjectComponent, IMissingBindingResolver
```
Comment (rprouse) [inline Source/Xamarin/Prism.Ninject.Forms/Extensions/DependencyServiceBindingResolver.cs]: Handles resolving from DependencyService when Ninject does not find the binding.

File: Source/Xamarin/Prism.Ninject.Forms/Extensions/DependencyServiceProvider.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        /// <inheritDoc />
+        public Type Type { get; private set; }
+
+        /// <inheritDoc />
+        public object Create(IContext context)
+        {
+            return ResolveFromDependencyService(Type);
+        }
+
+        private static object ResolveFromDependencyService(Type targetType)
```
Comment (rprouse) [inline Source/Xamarin/Prism.Ninject.Forms/Extensions/DependencyServiceProvider.cs]: Does the actual object creation from DependencyService. Code is cribbed from the Unity version.

File: Source/Xamarin/Prism.Ninject.Forms/NinjectBootstrapper.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+using Prism.Services;
+using Xamarin.Forms;
+using DependencyService = Prism.Services.DependencyService;
+
+namespace Prism.Ninject
+{
+    /// <summary>
+    /// A Ninject specific bootstrap class
+    /// </summary>
+    public abstract class NinjectBootstrapper : Bootstrapper
```
Comment (rprouse) [inline Source/Xamarin/Prism.Ninject.Forms/NinjectBootstrapper.cs]: This class is a line by line conversion of the Unity version.

File: Source/nuspecs/Prism.Ninject.nuspec [hunk 9]
```diff
@@ -22,10 +22,33 @@
         <dependency id="CommonServiceLocator" version="1.3.0" />
         <dependency id="Prism.Wpf" version="$wpfVersion$" />
       </group>
+
+      <group targetFramework="WindowsPhone8.0">
+        <dependency id="Prism.Forms" version="$formsVersion$" />
+        <dependency id="CommonServiceLocator" version="1.3" />
+        <dependency id="Xamarin.Forms" version="1.5.0.6447" />
+        <dependency id="Portable.Ninject" version="3.3.1" />
```
Comment (rprouse) [inline Source/nuspecs/Prism.Ninject.nuspec]: The nuspec is cribbed from the Unity version. It adds a dependency on Portable.Ninject which was ported by Oren Novotny.

**Comment by brianlagunas** [comment]
This looks great.  I'm at the MVP Summit this week, but will take a closer look next week and merge.  Unless @bartlannoeye wants to check it out before then :)

**Comment by brianlagunas** [comment]
This is great! Thanks!

**Pull request merged**